exporter/awscloudwatchlogsexporter/exporter.go

```diff
@@ -0,0 +1,235 @@
+	// TODO(jbd): Relax this once CW Logs support ingest
+	// without sequence tokens.
+	e.seqTokenMu.Lock()
+	defer e.seqTokenMu.Unlock()
+
+	logEvents, dropped := logsToCWLogs(e.logger, ld)
+	if len(logEvents) == 0 {
+		return 0, nil
+	}
+	if dropped > 0 {
+		e.logger.Warn("CloudWatch Logs exporter dropped log records", zap.Any("count", dropped))
+	}
+
+	e.logger.Debug("Putting log events", zap.Int("num_of_events", len(logEvents)))
+	input := &cloudwatchlogs.PutLogEventsInput{
+		LogGroupName:  aws.String(e.config.LogGroupName),
+		LogStreamName: aws.String(e.config.LogStreamName),
+		LogEvents:     logEvents,
+		SequenceToken: aws.String(e.seqToken),
+	}
+	out, err := e.client.PutLogEvents(input)
+	if err != nil {
+		return 0, err
+	}
+	if info := out.RejectedLogEventsInfo; info != nil {
+		return ld.LogRecordCount(), fmt.Errorf("log event rejected")
+	}
+	e.logger.Debug("Log events are successfully put")
+
+	e.seqToken = *out.NextSequenceToken
+	return 0, nil
+}
+
+func logsToCWLogs(logger *zap.Logger, ld pdata.Logs) ([]*cloudwatchlogs.InputLogEvent, int) {
+	n := ld.ResourceLogs().Len()
+	if n == 0 {
+		return []*cloudwatchlogs.InputLogEvent{}, 0
+	}
+
+	var dropped int
+	out := make([]*cloudwatchlogs.InputLogEvent, 0) // TODO(jbd): set a better capacity
+
+	rls := ld.ResourceLogs()
+	for i := 0; i < rls.Len(); i++ {
+		rl := rls.At(i)
+		resourceAttrs := attrsValue(rl.Resource().Attributes())
+
+		ills := rl.InstrumentationLibraryLogs()
+		for j := 0; j < ills.Len(); j++ {
+			ils := ills.At(j)
+			logs := ils.Logs()
+			for k := 0; k < logs.Len(); k++ {
+				log := logs.At(k)
+				event, err := logToCWLog(resourceAttrs, log)
+				if err != nil {
+					logger.Debug("Failed to convert to CloudWatch Log", zap.Error(err))
+					dropped++
+				} else {
+					out = append(out, event)
+				}
+			}
+		}
+	}
+	return out, dropped
+}
+
+type cwLogBody struct {
+	Name                   string                 `json:"name,omitempty"`
+	Body                   interface{}            `json:"body,omitempty"`
+	SeverityNumber         int32                  `json:"severity_number,omitempty"`
+	SeverityText           string                 `json:"severity_text,omitempty"`
+	DroppedAttributesCount uint32                 `json:"dropped_attributes_count,omitempty"`
+	Flags                  uint32                 `json:"flags,omitempty"`
+	TraceID                string                 `json:"trace_id,omitempty"`
+	SpanID                 string                 `json:"span_id,omitempty"`
+	Attributes             map[string]interface{} `json:"attributes,omitempty"`
+	Resource               map[string]interface{} `json:"resource,omitempty"`
+}
+
+func logToCWLog(resourceAttrs map[string]interface{}, log pdata.LogRecord) (*cloudwatchlogs.InputLogEvent, error) {
+	// TODO(jbd): Benchmark and improve the allocations.
+	// Evaluate go.elastic.co/fastjson as a replacement for encoding/json.
+	body := &cwLogBody{
```

Comment:
Nit: doesn't need to be a pointer. Probably doesn't matter anyway, hopefully Go's escape analysis can prove it can be stack allocated.